sdks/python/apache_beam/runners/portability/fn_api_runner/translations.py

```diff
@@ -704,6 +705,64 @@ def fix_side_input_pcoll_coders(stages, pipeline_context):
   return stages
 
 
+def eliminate_common_key_with_none(stages, context):
+  # type: (Iterable[Stage], TransformContext) -> Iterable[Stage]
+
+  """Runs common subexpression elimination for sibling KeyWithNone stages.
+
+  If multiple KeyWithNone stages share a common input, then all but one stages
+  will be eliminated along with their output PCollections. Transforms that
+  read input from the output of the eliminated KeyWithNone stages will be
+  remapped to read input from the output of the remaining KeyWithNone stage.
+  """
+  # Partition stages by whether they are eligible for common KeyWithNone
+  # elimination, and group eligible KeyWithNone stages by parent and
+  # environment.
+  grouped_eligible_stages = collections.defaultdict(list)
+  ineligible_stages = []
+  for stage in stages:
+    is_eligible = False
```

Comment:
Most of this is very general. What if you had a `merge_key(stage)` helper function which only returned something for this case (but was generally extensible) and then threw stuff into the dict according to this key (say, if it was not None), and then the rest of the logic could follow.

Reply:
Refactored the code to separate the key function from the grouping function, and used the grouping function in combiner packing as well; PTAL.